### ISSUE TITLE
Inconsistent behavior if fill data outside of array boundaries

### ISSUE THREAD

**Issue by VukW** [numpy/numpy#15495]
Sorry, if it is duplicate, but I didn't succeed to formulate search query properly.

When inserting data out of array boundaries, sometimes numpy don't  throw an error

### Reproducing code example:

```python
import numpy as np
a =  np.array([1,2,3])
a[10:12] = [10, 10]   # raises a ValueError
a[10:11] = [10]   # does nothing, no exception, just silently passes
```

I expect both lines behave consistently, f.ex. throw the same ValueError exception

### Numpy/Python version information:
numpy 1.18.1
Python 3.6.4 (v3.6.4:d48ecebad5, Dec 18 2017, 21:07:28) 
[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)]

**Comment by eric-wieser** [comment]
What you're seeing happen here is broadcasting - the failure is the same as:
```python
empty = np.arange(0)
empty[:] = [0]  # ok, arrays of length 1 broadcast
empty[:] = [0, 1]  # error
```
and
```python
nonempty = np.arange(4)
nonempty[:] = [0]  # ok, arrays of length 1 broadcast
nonempty[:] = [0, 1]  # error
```

**Comment by VukW** [comment]
Yeah, thanks!
I am confused with zero-size of the destination array. I always thought about broadcasting as "expanding" from 1 to N, not as "squeezing" to zero:). Also [docs](https://docs.scipy.org/doc/numpy/user/basics.broadcasting.html) list just expanding cases. Maybe it is worth to note here

**Comment by eric-wieser** [comment]
You're right, that page could do with an example with a length-0 dimension.

The key to understanding is that when "expanding" from 1 to N, `0` is a legal value for `N`.

**Comment by rossbar** [comment]
It's also worth noting that the reason that broadcasting comes into play is because slicing "outside of the array boundaries" doesn't result in an error, just an empty array:

```python
>>> a = np.array([1, 2, 3])
>>> b = a[10:12]
>>> b
array([], dtype=int64)
```

Then the [broadcasting rules](https://numpy.org/devdocs/user/basics.broadcasting.html?highlight=broadcasting#general-broadcasting-rules) come into play. So the expectation of getting an error for attempting to access data outside of the array contributes to the confusion.

I can't come up with a compelling broadcasting example for length-0 arrays (aside from slices resulting in single values), but if anyone came up with one I agree that it would be worthwhile to include in the documentation.

**Comment by eric-wieser** [comment]
> I can't come up with a compelling broadcasting example for length-0 arrays (aside from slices resulting in single values)

Perhaps:

```python
labels = np.array([1, 2, 3])
points = np.array([[1, 10], [2, 20], [3, 30]])
# this works whether `labels` contains 0 or not, due to the broadcasting rules also applying to length-0 arrays.
points[labels == 0] = [0, 100]   
```

**Comment by grudloff** [comment]
> You're right, that page could do with an example with a length-0 dimension.
> 
> The key to understanding is that when "expanding" from 1 to N, `0` is a legal value for `N` just like any other value.

I am sorry if I am not seeing something obvious here but why would a broadcast to zero be desired?

**Comment by eric-wieser** [comment]
Doesn't my comment above cover that? In a situation where the code is "broadcast this value to all elements satisfying this condition", you want the broadcast to succeed even if sometimes no elements satisfy the condition

**Comment by grudloff** [comment]
Oh I see! That makes it very clear, Thanks! Adding that example to the docs would be very helpful to understand this chosen behavior.

**Comment by eric-wieser** [comment]
As someone who went from not understanding to understanding clearly, do you want to take a stab at updating the docs yourself?
